lib/galaxy/schema/invocation.py

```diff
@@ -214,3 +231,114 @@ class InvocationMessageResponseModel(BaseModel):
+    SCHEDULED = "scheduled"  # Workflow invocation step has been scheduled.
+    # CANCELLED = 'cancelled',  TODO: implement and expose
+    # FAILED = 'failed',  TODO: implement and expose
+
+
+class InvocationStepOutput(Model):
+    src: str = Field(
+        literal_to_value(INVOCATION_STEP_OUTPUT_SRC),
+        title="src",
+        description="The source model of the output.",
+        const=True,
+        mark_required_in_schema=True,
+    )
+    id: DecodedDatabaseIdField = Field(
+        ...,
+        title="Dataset ID",
+        description="Dataset ID of the workflow step output.",
+    )
+    uuid: Optional[schema.UUID4] = Field(
+        None,
+        title="UUID",
+        description="Universal unique identifier of the workflow step output dataset.",
+    )
+
+
+class InvocationStepCollectionOutput(Model):
+    src: str = Field(
+        literal_to_value(INVOCATION_STEP_COLLECTION_OUTPUT_SRC),
+        title="src",
+        description="The source model of the output.",
+        const=True,
+        mark_required_in_schema=True,
+    )
+    id: DecodedDatabaseIdField = Field(
+        ...,
+        title="Dataset Collection ID",
+        description="Dataset Collection ID of the workflow step output.",
+    )
+
+
+class InvocationStep(Model):
+    """Information about Workflow Invocation Step"""
+
+    model_class: schema.INVOCATION_STEP_MODEL_CLASS = schema.ModelClassField(schema.INVOCATION_STEP_MODEL_CLASS)
+    id: DecodedDatabaseIdField = schema.EntityIdField
+    update_time: Optional[datetime] = schema.UpdateTimeField
+    job_id: Optional[DecodedDatabaseIdField] = Field(
+        title="Job ID",
+        description="The encoded ID of the job associated with this workflow invocation step.",
+    )
+    workflow_step_id: DecodedDatabaseIdField = Field(
+        ...,
+        title="Workflow step ID",
+        description="The encoded ID of the workflow step associated with this workflow invocation step.",
+    )
+    subworkflow_invocation_id: DecodedDatabaseIdField = Field(
+        title="Subworkflow invocation ID",
+        description="The encoded ID of the subworkflow invocation.",
+    )
+    state: InvocationStepState = Field(
+        ...,
+        title="State of the invocation step",
+        description="Describes where in the scheduling process the workflow invocation step is.",
+    )
+    action: bool = InvocationStepAction
+    order_index: int = Field(
+        ...,
+        title="Order index",
+        description="The index of the workflow step in the workflow.",
+    )
+    workflow_step_label: str = Field(
+        title="Step label",
+        description="The label of the workflow step",
+    )
+    workflow_step_uuid: Optional[schema.UUID4] = Field(
+        None,
+        title="UUID",
+        description="Universal unique identifier of the workflow step.",
+    )
+    outputs: Dict[str, InvocationStepOutput] = Field(
+        {},
```

Comment:
This default value might be correct (I haven't checked) but it is important to note that what makes a field `optional` in FastAPI is not using `Optional[whatever]` in the type hint (this just helps the type checker), but providing a default value.

Again, this might be correct in this particular model, but I was confused in the past by this fact so I wanted to clarify here.

The good news is that now we can easily check this by looking at the generated client schema:
```ts
/**
* Outputs
* @description The outputs of the workflow invocation step.
* @default {}
*/
outputs?: {
  [key: string]: components["schemas"]["InvocationStepOutput"] | undefined;
};
```

Observe that `outputs` can be `undefined`.

Reply:
The explicit `optional` vs (non)present default value is surely a bit confusing, but I am not sure it applies here. IIUC the API will return an empty dict if there are no outputs -- see https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/model/__init__.py#L8361